File: src/net/rizon/moo/servermonitor/certChecker.scala
package net.rizon.moo.servermonitor

import net.rizon.moo._

import java.util.Date
import java.util.Calendar
import java.util.HashSet
import java.security.cert._

object certChecker
{
	var sslPort = 6697
	var warned = new HashSet[String]
	var error = new HashSet[String]
	var lastIdx = -1
	
	private def report(what: String)
	{
		for (chan <- moo.conf.getAdminChannels())
			moo.reply(null, chan, "[CERTCHECKER] " + what)
	}
	
	def run()
	{
		var servers = server.getServers()
		lastIdx += 1
		if (lastIdx >= servers.length)
			lastIdx = 0
		for (lastIdx <- lastIdx to servers.length)
		{
			var server = servers(lastIdx)
			var ok = true
			
			if (!server.isHub() && !server.isServices())
			{
				if (server.cert == null)
				{
					var sc = new scheck(server, moo.conf.getAdminChannels(), true, certChecker.sslPort, true, false);
					sc.start
					return
				}
				else
				{
					try
					{
						var now = new Date()
						server.cert.checkValidity(now)
						
						try
						{
							var c = Calendar.getInstance()
							c.setTime(now)
							c.add(Calendar.DATE, 1)
							server.cert.checkValidity(c.getTime())
						}
						catch
						{
							case e: CertificateExpiredException =>
							{
								ok = false
								if (!certChecker.warned.contains(server.getName()) && !certChecker.error.contains(server.getName()))
								{
									report("SSL certificate for " + server.getName() + " expires on " + server.cert.getNotAfter() + ", which is " + moo.difference(now, server.cert.getNotAfter()) + " from now")
									certChecker.warned.add(server.getName())
								}
							}
						}
					}
					catch
					{
						case e: CertificateExpiredException =>
						{
							ok = false
							if (!certChecker.error.contains(server.getName()))
							{
								report(server.getName + " has EXPIRED X509 SSL certificate!");
								certChecker.error.add(server.getName())
							}
						}
						case e: CertificateNotYetValidException =>
						{
							ok = false
							if (!certChecker.error.contains(server.getName()))
							{
								report(server.getName + " has a NOT VALID YET X509 SSL certificate!");
								certChecker.error.add(server.getName())
							}
						}
					}
					
					if (ok)
					{
						certChecker.warned.remove(server.getName())
						certChecker.error.remove(server.getName())
					}
				}
			}
		}
	}
}

File: servermonitor/src/main/java/net/rizon/moo/plugin/servermonitor/CertChecker.java
package net.rizon.moo.plugin.servermonitor;

import com.google.inject.Inject;
import net.rizon.moo.Moo;
import net.rizon.moo.conf.Config;
import net.rizon.moo.irc.Server;
import net.rizon.moo.irc.ServerManager;
import net.rizon.moo.plugin.servermonitor.scheck.SCheck;

class CertChecker implements Runnable
{
	private static final int sslPort = 6697;
	private static int lastIdx = -1;
	
	@Inject
	private ServerManager serverManager;
	
	@Inject
	private Config conf;

	@Override
	public void run()
	{
		Server[] servers = serverManager.getServers();
		if (servers.length == 0)
			return;

		lastIdx += 1;
		if (lastIdx >= servers.length)
			lastIdx = 0;

		Server s = servers[lastIdx];

		if (!s.isHub() && !s.isServices())
		{
			SCheck sc = new SCheck(s, conf.moo_log_channels, true, CertChecker.sslPort, true, false);
			Moo.injector.injectMembers(sc);
			sc.start();
		}
	}
}